### PR TITLE
Lua

### PR DESCRIPTION
**Pull request by Doralitze**
This adds the Lua filter as well as realted bug fixes.

### CODE REVIEW

File: src/filters/filter_lua_script.cpp [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+        lua.set_function( "hsi_to_rgb", sol::overload(
+                hsi_to_rgb_color,
+                hsi_to_rgb_table
+        ) );
+
+        lua.set_function( "hsi_to_rgbw", sol::overload(
+                hsi_to_rgbw_color,
+                hsi_to_rgbw_table
+        ) );
```
Comment (Doralitze) [inline src/filters/filter_lua_script.cpp:230]: As we should out source the methods of the API, so should we do for context creation as we might need this behaviour in multiple places as well.

File: src/filters/filter_lua_script.cpp [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+        try {
+            lua.script(initial_parameters.at("script"));
+        } catch (std::exception &e){
+            throw filter_config_exception(std::string("setup the filter threw an error: ") + e.what());
+        }
+
+        sol::object scene_activated_obj = lua["scene_activated"];
```
Comment (Doralitze) [inline src/filters/filter_lua_script.cpp:238]: What happens if the User does not provide this method? Will the filter crash?

File: src/filters/filter_lua_script.cpp [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        sol::object update_obj = lua["update"];
+        if (update_obj.get_type() == sol::type::function && update_obj.is<std::function<void()>>()){
+            update_lua = update_obj;
+        } else {
+            throw filter_config_exception("update is not a function or has the wrong signature");
+        }
+    }
+
+    bool filter_lua_script::receive_update_from_gui(const std::string &key, const std::string &_value) {
+
```
Comment (Doralitze) [inline src/filters/filter_lua_script.cpp:254]: We should also check if we can forward the k,v pair to the script and return the processing result from the script. If the corresponding method isn't implemented we should return false right away.

File: src/filters/filter_lua_script.cpp [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+                            if (channel.get_type() == sol::type::number) {
+                                uint8_t value = ((sol::table) universe)[chan];
+                                (*uptr)[chan] = value;
+//                                std::cout << "test : " << (int) value << std::endl;
+//                            } else if (channel.get_type() != sol::type::nil) {
+//                                std::cout << " test : is not nil : " << std::endl;
+                            }
+                        }
+                    }
+                }
```
Comment (Doralitze) [inline src/filters/filter_lua_script.cpp:340]: If we have large code blocks doing just one isolated thing we should introduce an inlined method for that.

File: test/scheduler_test.ldflags
Comment (Doralitze) [inline test/scheduler_test.ldflags:1]: Why was this file deleted? Don't we need it any more?

**Review by Doralitze — COMMENTED** [review]
@LRahlff I read through your code and made some comments.

File: README.md [hunk 5]
```diff
@@ -13,6 +13,8 @@ The part of the software that translates the users wishes to DMX in real time.
 	* libusb-dev
 	* libftdi-dev
 	* pkg-config
+	* liblua5.4-dev
+    * lua5.4
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply README.md:17]: removed

File: test/scheduler_test.ldflags
Comment (LRahlff) [reply test/scheduler_test.ldflags:1]: do we need these ldflags files? i thought these are only made by executing make. (and dont get deleted if something went wrong/user interuption)

File: src/filters/filter_conversion.hpp [hunk 6]
```diff
@@ -299,47 +300,137 @@ COMPILER_SUPRESS("-Weffc++")
                 g = 0;
                 w = (uint8_t) std::round(255*(1-S)*I);
             }
+//            input->pixel_to_rgb(r, g, b, w);
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/filters/filter_conversion.hpp:303]: done, added testcases

File: src/filters/filter_conversion.hpp [hunk 7]
```diff
@@ -299,47 +300,137 @@ COMPILER_SUPRESS("-Weffc++")
...
+        }
+
+        virtual void get_output_channels(channel_mapping& map, const std::string& name) override {
+            map.float_channels[name + ":hue"] = &h;
+            map.float_channels[name + ":saturation"] = &s;
+            map.float_channels[name + ":iluminance"] = &i;
+        }
+
+        virtual void update() override {
+            this->h = (*this->input).hue;
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/filters/filter_conversion.hpp:379]: done

File: src/filters/filter_lua_script.cpp (hunk 2 above)
Comment (LRahlff) [reply src/filters/filter_lua_script.cpp:238]: Yes or no: if script is not privided the filter will throw an error. if scene_activated is not available it will throw another error.
If the user does not do anything with scene_activated, then the function of line 213 is used.

File: src/filters/filter_cue.cpp [hunk 8]
```diff
@@ -18,6 +19,53 @@ int count_occurence_of(const std::string &base_string, std::string pattern, size
...
+            channel_names_float.push_back(channel_name);
+            float_channels.push_back(0);
+            last_float_channels.push_back(0);
+        } else {
+            channel.push_back(channel_str(COLOR, color_channels.size()));
+            channel_names_color.push_back(channel_name);
+            color_channels.push_back(dmxfish::dmx::pixel());
+            last_color_channels.push_back(dmxfish::dmx::pixel());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/filters/filter_cue.cpp:67]: yes, but then we need to count the nr of this (channel within one type) in another way

File: src/filters/filter_fader.hpp [hunk 9]
```diff
@@ -12,7 +12,9 @@
 #include "lib/macros.hpp"
 #include "main.hpp"
 
+#include "lib/logging.hpp"
 #include "control_desk/bank_column.hpp"
+#include "lib/logging.hpp"
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/filters/filter_fader.hpp:17]: removed

File: src/filters/filter_lua_script.cpp (hunk 3 above)
Comment (LRahlff) [reply src/filters/filter_lua_script.cpp:254]: a key value pair for each method, set by the user? Is it not enough to throw an error if something in update goes wrong? which type of exception can the caller of update handle?

File: src/filters/filter_lua_script.cpp [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+//                            sol::object ch_name = channel_kv_pair.first;
+//                            sol::object ch_value = channel_kv_pair.second;
+//                            (*uptr)[ch_name] = ch_value;
+//                            std::cout << "c " << std::endl;
+//                        }
+//                    } else {
+//                            throw std::invalid_argument("The requested universe with id " + std::to_string(uni_key) +
+//                                                        " does not exist anymore");
+//                    }
+//                }
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/filters/filter_lua_script.cpp:368]: these were thoughts about getting the values with the patching informations

File: src/filters/filter_lua_script.cpp (hunk 4 above)
Comment (LRahlff) [reply src/filters/filter_lua_script.cpp:340]: moved

**Comment by Doralitze** [comment]
No, the ldflags files are read by the make file in order to determine the linker flags for the corresponding executables.

File: src/filters/filter_cue.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Doralitze) [reply src/filters/filter_cue.cpp:67]: OK, if this fits this purpose, we shall leave it as is.

File: src/filters/filter_lua_script.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Doralitze) [reply src/filters/filter_lua_script.cpp]: Well I still think that it would be better to perform that lookup in lua as we then do not have to do it on every update. This being done in Lua would also provide more flexibility (for example as we did at Erstiparty).

File: src/filters/filter_lua_script.cpp (hunk 2 above)
Comment (Doralitze) [reply src/filters/filter_lua_script.cpp:284]: OK, we should make sure then, that this error will not crash fish and is being quite informative (as described with issue 78 of the editor).

File: src/filters/filter_lua_script.cpp (hunk 3 above)
Comment (Doralitze) [reply src/filters/filter_lua_script.cpp:300]: No, you got me wrong. We should do something like this pseudo-code:
```
bool receive_update_from_gui(k, v):
    if not lua_context.has_function("receive_gui_update"):
        return false
    else:
        try:
            return (bool) lua_context.exec_function("receive_gui_update", k, v)
        catch:
            return false
```

File: test/scheduler_test.ldflags
Comment (Doralitze) [reply test/scheduler_test.ldflags:1]: No, the ldflags files are read by the make file in order to determine the linker flags for the corresponding executables.

File: src/control_desk/bank_column.cpp [hunk 11]
```diff
@@ -101,22 +105,48 @@ namespace dmxfish::control_desk {
...
 		update_physical_fader_position();
+        if (current_bank_mode == bank_mode::DIRECT_INPUT_MODE) {
+            update_display_text();
+        }
 		send_col_update_to_fish();
 	}
 
 	void bank_column::process_encoder_change_message(int change_request) {
-		raw_configuration.rotary_position += (int16_t) change_request;
+        this->accumulated_change++;
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/control_desk/bank_column.cpp:134]: is it just, if process_encoder_change_message gets called more often, the the change value gets multiplied by a higher value (which increases each time by 1)? How big can change_request can get? But limiting this sounds great.

File: src/filters/filter_lua_script.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/filters/filter_lua_script.cpp]: Why we dont have to do that on every update? Now it has to be done in lua on every update, or do i think wrong?
If we use a struct that lua knows, then lua writes directly in this memory and we have to update each channel as well, but dont have to lookup these values in lua.

File: test/scheduler_test.ldflags
Comment (LRahlff) [reply test/scheduler_test.ldflags:1]: so it can stay deleted or did you wrote something in there?

File: src/filters/filter_cue.cpp [hunk 12]
```diff
@@ -647,6 +650,14 @@ namespace dmxfish::filters {
...
 
     void filter_cue::scene_activated() {
+	    if (this->default_cue > -1) {
+		    this->active_cue = (uint16_t) this->default_cue + 1;
+		    update_last_values();
+                    start_time = *time;
+                    last_timestamp = *time;
+                    frame = 0;
+		    ::spdlog::info("Switched to Cue {}.", this->default_cue + 1);
+	    }
```
[the remark this answers is not in the document]
Comment (LRahlff) [reply src/filters/filter_cue.cpp:635]: i tried to do that, but im not sure if everything is correct right now

File: src/filters/filter_lua_script.cpp (hunk 3 above)
Comment (LRahlff) [reply src/filters/filter_lua_script.cpp]: should work right now

File: src/filters/filter_cue.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (Doralitze) [reply src/filters/filter_cue.cpp:639]: Well we should test it then.

**Review by Doralitze — COMMENTED** [review]
We should test these changes soon and then continue with the merge.

File: test/scheduler_test.ldflags
Comment (Doralitze) [reply test/scheduler_test.ldflags:1]: I think it can stay deleted.

File: src/control_desk/bank_column.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Doralitze) [reply src/control_desk/bank_column.cpp]: Exactly. The value will also be reset to 0 after a second without dialing the knob. This way one can enter values quicker while also preserving the accuracy. It is limited to 1024 now.

File: src/control_desk/bank_column.cpp [hunk 13]
```diff
@@ -247,14 +282,17 @@ namespace dmxfish::control_desk {
...
 					const auto tpos = i + display_scroll_position_up;
 					if(tpos < up_text.length()) {
 						content[i] = up_text.at(tpos);
 					} else {
 						content[i] = ' ';
 					}
 				}
+                if (current_bank_mode == bank_mode::DIRECT_INPUT_MODE) {
+                    content[6] = (raw_working_on_primary) ? '+' : '*';
+                }
```
[the remark this answers is not in the document]
Comment (Doralitze) [reply src/control_desk/bank_column.cpp]: Done

File: src/filters/filter_lua_functions.hpp [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+ */
+COMPILER_SUPRESS("-Wswitch-default")
+COMPILER_SUPRESS("-Weffc++")
+COMPILER_SUPRESS("-Wuseless-cast")
+COMPILER_SUPRESS("-Wswitch-enum")
+#include <sol/sol.hpp>
+COMPILER_RESTORE("-Wswitch-enum")
+COMPILER_RESTORE("-Wuseless-cast")
+COMPILER_RESTORE("-Weffc++")
+COMPILER_RESTORE("-Wswitch-default")
```
Comment (LRahlff) [inline src/filters/filter_lua_functions.hpp:14]: how to suppress warnings of an external library (only header)?

Comment (Doralitze) [reply src/filters/filter_lua_functions.hpp:14]: Well, you can disable warnings completely but I would't recommend that, as the warnings above are due to poor coding style within `sol` but it could also throw warnings if the library is used the wrong way.

**Review by Doralitze — COMMENTED** [review]
I think we're fine now

File: src/filters/filter_lua_script.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Doralitze) [reply src/filters/filter_lua_script.cpp]: Yeah, I think that you're right about that. This loop will also do not do anything as long as there was no output written. I would have put it inside a separate method for readability anyway.

**Review by LRahlff — APPROVED** [review]
For me it looks fine too

File: src/filters/filter_lua_script.cpp (hunk 2 above)
Comment (Doralitze) [reply src/filters/filter_lua_script.cpp:258]: Do we have a test for this behavior now?

Comment (Doralitze) [reply src/filters/filter_lua_script.cpp:258]: If not: I'll write one real quick.

Comment (Doralitze) [reply src/filters/filter_lua_script.cpp:258]: Never mind, the test is there. I'll merge now.

Comment (LRahlff) [reply src/filters/filter_lua_script.cpp:258]: added a little more information to the message
